packages/matter-node-ble.js/src/ble/NobleBleClient.ts

```diff
@@ -0,0 +1,103 @@
+import { ByteArray } from "@project-chip/matter.js/util";
+import { BLE_MATTER_SERVICE_UUID } from "@project-chip/matter.js/ble";
+import { Logger } from "@project-chip/matter.js/log";
+
+const logger = Logger.get("NobleBleClient");
+let noble: typeof import("@abandonware/noble");
+
+function loadNoble(hciId = 0) {
+    // load noble driver with the correct device selected
+    process.env.NOBLE_HCI_DEVICE_ID = hciId.toString();
+    try {
+        noble = require("@abandonware/noble");
+        if (typeof noble.on !== "function") {
+            // The following commit broke the default exported instance of noble:
+            // https://github.com/abandonware/noble/commit/b67eea246f719947fc45b1b52b856e61637a8a8e
+            noble = (noble as any)({ extended: false });
+        }
+    } catch (error: any) {
+        logger.error(`Error loading noble: ${error.message}`);
+        throw error;
+    }
+}
+
+export class NobleBleClient {
+    private readonly discoveredPeripherals = new Map<string, { peripheral: Peripheral, matterServiceData: ByteArray }>();
+    private shouldScan = false;
+    private isScanning = false;
+    private nobleState = 'unknown';
+    private deviceDiscoveredCallback: ((peripheral: Peripheral, manufacturerData: ByteArray) => void) | undefined;
+
+    constructor() {
+        loadNoble();
+        noble.on('stateChange', state => {
+            this.nobleState = state;
+            logger.debug(`Noble state changed to ${state}`);
+            if (state === 'poweredOn') {
+                if (this.shouldScan) {
+                    void this.startScanning();
+                }
+            } else {
+                void this.stopScanning();
+            }
+        });
+        noble.on('discover', peripheral => {
+            void this.handleDiscoveredDevice(peripheral)
+        });
+        noble.on('scanStart', () => this.isScanning = true);
+        noble.on('scanStop', () => this.isScanning = false);
+    }
+
+    public setDiscoveryCallback(callback: (peripheral: Peripheral, manufacturerData: ByteArray) => void) {
+        this.deviceDiscoveredCallback = callback;
+        for (const { peripheral, matterServiceData } of this.discoveredPeripherals.values()) {
+            this.deviceDiscoveredCallback(peripheral, matterServiceData);
+        }
+    }
+
+    public async startScanning() {
+        if (this.isScanning) return;
+
+        this.shouldScan = true;
+        if (this.nobleState === 'poweredOn') {
+            logger.debug('Start scanning for Matter Services ...');
+            await noble.startScanningAsync([BLE_MATTER_SERVICE_UUID], false);
+        } else {
+            logger.debug('noble state is not poweredOn ... delay scanning till poweredOn');
+        }
+    }
+
+    public async stopScanning() {
+        this.shouldScan = false;
+        await noble.stopScanningAsync();
+    }
+
+    private async handleDiscoveredDevice(peripheral: Peripheral) {
```

Comment:
Probably this isn't supposed to be async?

Reply:
right. fixed